### PR TITLE
Added networking example to listen on random socket using tcp/ip

### LINKED ISSUE #123: Get a Random Unused TCP/IP or UDP Port

### PR DESCRIPTION
**Pull request by AndyGauge**
Fixes #123

### CODE REVIEW

File: src/net.md [hunk 1]
```diff
@@ -537,6 +538,73 @@ fn run() -> Result<()> {
...
+}
+
+fn run() ->  Result<()>{
+    let loopback = Ipv4Addr::new(127,0,0,1);
+    // Assigning port 0 requests the OS to assign a free port
+    let socket = SocketAddrV4::new(loopback, 0);
+    let listen = TcpListener::bind(socket)?;
+    let port = listen.local_addr()?;
+    println!("Listening on {}, access this port to end the program", port);
+    match listen.accept() {
```
Comment (dtolnay) [inline src/net.md]: Please use `listen.accept()?` to handle this error.

File: src/net.md [hunk 2]
```diff
@@ -537,6 +538,73 @@ fn run() -> Result<()> {
...
+        Ok((mut tcp_stream, addr)) => {
+            // read from the socket until connection closed by client.
+            let mut input = String::new();
+            let _ = tcp_stream.read_to_string(&mut input); //discard byte count
```
Comment (dtolnay) [inline src/net.md]: This can return an error, please handle it with `?`.

File: src/net.md [hunk 3]
```diff
@@ -537,6 +538,73 @@ fn run() -> Result<()> {
...
+
+By passing 0 to the [`TcpListener::bind`], the OS will assign an unused random
+port.  The address and port that the OS assigns is available using
+[`TcpListener::local_addr`].
+
+The rest of the recipe prints out the request made on the socket.
+`TcpListener::accept` returns a tuple of `TcpStream` and client information
+within a `Result`.  The `Read` implmentation is used on that `TcpStream` to
+collect the request payload.  Closing the program is as easy as browsing to the
+ip:port or using telnet to send some data, pressing ctrl-] and typing quit.
```
Comment (dtolnay) [inline src/net.md]: Please give the telnet command to run here so that users can tests this.

File: src/net.md [hunk 4]
```diff
@@ -537,6 +538,73 @@ fn run() -> Result<()> {
...
+use std::net::{SocketAddrV4, Ipv4Addr,TcpListener};
+use std::io::Read;
+
+error_chain! {
+    foreign_links {
+        Io(::std::io::Error);
+    }
+}
+
+fn run() ->  Result<()>{
```
Comment (dtolnay) [inline src/net.md]: There should be one space before `Result<()>` and one space after.

File: src/net.md [hunk 5]
```diff
@@ -537,6 +538,73 @@ fn run() -> Result<()> {
...
+request is made.  If you use your browser to test this program, simply enter
+the address:port into your location bar and make the request.  Because the
+program returns nothing, the browser's stop feature can be used to speed things
+up.
+
+```rust, no_run
+#[macro_use]
+extern crate error_chain;
+
+use std::net::{SocketAddrV4, Ipv4Addr,TcpListener};
```
Comment (dtolnay) [inline src/net.md]: Please put one space after both commas.

File: src/net.md [hunk 6]
```diff
@@ -537,6 +538,73 @@ fn run() -> Result<()> {
...
+use std::io::Read;
+
+error_chain! {
+    foreign_links {
+        Io(::std::io::Error);
+    }
+}
+
+fn run() ->  Result<()>{
+    let loopback = Ipv4Addr::new(127,0,0,1);
```
Comment (dtolnay) [inline src/net.md]: Rust style is to put spaces after commas.

**Comment by dtolnay** [comment]
Hey @budziq could you take over the review for this one? LMK when it's ready to merge. Thanks so much!

File: src/net.md (hunk 1 above)
Comment (AndyGauge) [inline src/net.md]: Is it idiomatic to keep the match?  I've refactored it to this:
```rust
    match listen.accept()? {
        (mut tcp_stream, addr) => {
            // read from the socket until connection closed by client.
            let mut input = String::new();
            let _ = tcp_stream.read_to_string(&mut input)?; //discard byte count
            println!("Connection received! \r\n{:?} says {}", addr, input);
            Ok(())
        }
    }
```

Comment (dtolnay) [inline src/net.md]: Would this work?

```rust
let (mut tcp_stream, addr) = listen.accept()?;
```

**Comment by budziq** [comment]
Will do. I'll be near a pc later today

File: src/net.md [hunk 7]
```diff
@@ -537,6 +538,71 @@ fn run() -> Result<()> {
...
+    foreign_links {
+        Io(::std::io::Error);
+    }
+}
+
+fn run() -> Result<()> {
+    let loopback = Ipv4Addr::new(127, 0, 0, 1);
+    // Assigning port 0 requests the OS to assign a free port
+    let socket = SocketAddrV4::new(loopback, 0);
+    let listen = TcpListener::bind(socket)?;
```
Comment (budziq) [inline src/net.md]: I would suggest renaming listen to listener

File: src/net.md [hunk 8]
```diff
@@ -537,6 +538,71 @@ fn run() -> Result<()> {
...
+    let port = listen.local_addr()?;
+    println!("Listening on {}, access this port to end the program", port);
+    let (mut tcp_stream, addr) = listen.accept()?;  //block  until requested
+    let mut input = String::new();
```
Comment (budziq) [inline src/net.md]: the connection is already accepted here. I would suggest printing 
`println!("Connection received on {:?}? says {}", addr);`
here as it shows exactly where synchronous waiting occurs.

File: src/net.md [hunk 9]
```diff
@@ -537,6 +538,71 @@ fn run() -> Result<()> {
...
+    // read from the socket until connection closed by client.
+    let _ = tcp_stream.read_to_string(&mut input)?; //discard byte count
+    println!("Connection received! \r\n{:?} says {}", addr, input);
```
Comment (budziq) [inline src/net.md]: I would drop the \r\n and just use
`println!("{:?} says {}", addr, input);`
here

File: src/net.md [hunk 10]
```diff
@@ -537,6 +538,71 @@ fn run() -> Result<()> {
...
+    let loopback = Ipv4Addr::new(127, 0, 0, 1);
+    // Assigning port 0 requests the OS to assign a free port
+    let socket = SocketAddrV4::new(loopback, 0);
+    let listen = TcpListener::bind(socket)?;
+    let port = listen.local_addr()?;
+    println!("Listening on {}, access this port to end the program", port);
+    let (mut tcp_stream, addr) = listen.accept()?;  //block  until requested
+    let mut input = String::new();
+    // read from the socket until connection closed by client.
+    let _ = tcp_stream.read_to_string(&mut input)?; //discard byte count
```
Comment (budziq) [inline src/net.md]: I would rephrase the comment above to include the 
`//discard byte count`

File: src/net.md [hunk 11]
```diff
@@ -537,6 +538,71 @@ fn run() -> Result<()> {
...
+    println!("Connection received! \r\n{:?} says {}", addr, input);
+    Ok(())
+}
+
+quick_main!(run);
+```
+
+The `std` library is leveraged to make a well formed IP/port with the
+`SocketAddrV4` and `Ipv4Addr` structs.  The loopback address is a special
```
Comment (budziq) [inline src/net.md]: Please add hyperlinks for `SocketAddrV4` and `Ipv4Addr`

File: src/net.md [hunk 12]
```diff
@@ -537,6 +538,71 @@ fn run() -> Result<()> {
...
+address that runs only on the local machine, and is available on all machines.
+
+By passing 0 to the [`TcpListener::bind`], the OS will assign an unused random
+port.  The address and port that the OS assigns is available using
+[`TcpListener::local_addr`].
+
+The rest of the recipe prints out the request made on the socket.
+`TcpListener::accept` returns a tuple of `TcpStream` and client information
```
Comment (budziq) [inline src/net.md]: Please add hyperlinks for `TcpListener::accept` and `TcpStream` 

also what "client information"? I would suggest just writing:
 `(TcpStream, SocketAddr)` with appropriate hyperlinks

File: src/net.md [hunk 13]
```diff
@@ -537,6 +538,71 @@ fn run() -> Result<()> {
...
+within a `Result`.  The `Read` implmentation is used on that `TcpStream` to
+collect the request payload.  Closing the program is as easy as browsing to the
+ip:port or using `telnet ip port` to send some data.  For example, if the
+program shows Listening on 127.0.0.1:11500, run `telnet 127.0.0.1 11500`.  After
```
Comment (budziq) [inline src/net.md]: I would suggest giving the  `telnet 127.0.0.1 11500` a separate line

File: src/net.md [hunk 14]
```diff
@@ -537,6 +538,71 @@ fn run() -> Result<()> {
...
+`SocketAddrV4` and `Ipv4Addr` structs.  The loopback address is a special
+address that runs only on the local machine, and is available on all machines.
+
+By passing 0 to the [`TcpListener::bind`], the OS will assign an unused random
+port.  The address and port that the OS assigns is available using
+[`TcpListener::local_addr`].
+
+The rest of the recipe prints out the request made on the socket.
+`TcpListener::accept` returns a tuple of `TcpStream` and client information
+within a `Result`.  The `Read` implmentation is used on that `TcpStream` to
```
Comment (budziq) [inline src/net.md]: I would suggest dropping the information about `Result` once hyperlinks are added to other structs and functions as it makes the description overly wordy

File: src/net.md (hunk 12 above)
Comment (budziq) [inline src/net.md]: I would suggest using passive voice and not describing what function returns what as reader will be able to just follow your hyperlinks. But rather write what is happening in a concise 2-4 sentence description.

File: src/net.md [hunk 15]
```diff
@@ -537,6 +538,71 @@ fn run() -> Result<()> {
...
+collect the request payload.  Closing the program is as easy as browsing to the
```
Comment (budziq) [inline src/net.md]: I would suggest rewording the description to. "Reading on the socket with `read_to_string` will wait untill connection is closed" which can be tested with `telnet ip port` command.

File: src/net.md [hunk 16]
```diff
@@ -537,6 +538,75 @@ fn run() -> Result<()> {
 quick_main!(run);
 ```
 
+[ex-random-port-tcp]: #ex-random-port-tcp
+<a name="ex-random-port-tcp"></a>
+## Listen on Unused port TCP/IP
+
+[![std-badge]][std] [![cat-net-badge]][cat-net]
+
+Give the OS the responsibility to pick an unused [registered port].  These ports
```
Comment (budziq) [inline src/net.md]: I know that you have put a lot  of effort into this section but I would rather not educate readers on CS/Networking basics. We can safely assume that reader understands these concepts and came here only for the rust snippet and its very brief textual description (that would ideally be 2-4 sentences). I would suggest dropping this paragraph completely.

File: src/net.md [hunk 17]
```diff
@@ -537,6 +538,75 @@ fn run() -> Result<()> {
...
+    println!("{:?} says {}", addr, input);
+    Ok(())
+}
+
+quick_main!(run);
+```
+
+The `std` library is leveraged to make a well formed IP/port with the
+[`SocketAddrV4`] and [`Ipv4Addr`] structs.  The loopback address is a special
+address that runs only on the local machine, and is available on all machines.
```
Comment (budziq) [inline src/net.md]: I would rather not try to educate the readers on what the loopback interface is. I would just stick to stating that code is listening on the loopback interface (possibly with hyperlink to some authoritative resource)

File: src/net.md [hunk 18]
```diff
@@ -537,6 +538,75 @@ fn run() -> Result<()> {
...
+generally do not require sudo permission to bind to, and can be helpful when
+establishing connections with clients in a disposable communication chain.
+These ports begin with 1024 and have around 48,000 available.  Typically, a
+known port is used to initiate the communication and subsequent communication
+occurs on the registered port after a handshake process.
+
+In this example, the port is displayed on the console, and will listen until a
+request is made.  If you use your browser to test this program, simply enter
+the address:port into your location bar and make the request.  Because the
+program returns nothing, the browser's stop feature can be used to speed things
```
Comment (budziq) [inline src/net.md]: This paragraph is overly word and redundant with later telnet testing one. 
I would remove all of it except the first sentence.

File: src/net.md [hunk 19]
```diff
@@ -537,6 +538,75 @@ fn run() -> Result<()> {
...
+
+The `std` library is leveraged to make a well formed IP/port with the
+[`SocketAddrV4`] and [`Ipv4Addr`] structs.  The loopback address is a special
+address that runs only on the local machine, and is available on all machines.
+
+By passing 0 to the [`TcpListener::bind`], the OS will assign an unused random
+port.  The address and port that the OS assigns is available using
+[`TcpListener::local_addr`].
+
+The rest of the recipe prints out the request made on the socket.
```
Comment (budziq) [inline src/net.md]: I would remove this sentence and modify the following ones to reflect what is going one step by step.

File: src/net.md [hunk 20]
```diff
@@ -537,6 +538,75 @@ fn run() -> Result<()> {
...
+}
+
+quick_main!(run);
+```
+
+The `std` library is leveraged to make a well formed IP/port with the
+[`SocketAddrV4`] and [`Ipv4Addr`] structs.  The loopback address is a special
+address that runs only on the local machine, and is available on all machines.
+
+By passing 0 to the [`TcpListener::bind`], the OS will assign an unused random
```
Comment (budziq) [inline src/net.md]: I would suggest rewording it in a more concise manner e.g:

> An unused random port is requested by passing 0 to the [`TcpListener::bind`]. The assigned address is available via [`TcpListener::local_addr`].

File: src/net.md [hunk 21]
```diff
@@ -537,6 +538,75 @@ fn run() -> Result<()> {
...
+port.  The address and port that the OS assigns is available using
+[`TcpListener::local_addr`].
+
+The rest of the recipe prints out the request made on the socket.
+[`TcpListener::accept`] returns a `(`[`TcpStream`],  [`SocketAddrV4`]`)`
```
Comment (budziq) [inline src/net.md]: The key fact about accept is that it **synchronously waits for an incoming** connection and returns the pair

File: src/net.md [hunk 22]
```diff
@@ -537,6 +538,75 @@ fn run() -> Result<()> {
...
+representing the data from the client, its IP address and port.  Reading on the
```
Comment (budziq) [inline src/net.md]: >data from the client, its IP address and port.
Could be misunderstood. I would suggest something like:
"representing client connection IP address and port."

File: src/net.md [hunk 23]
```diff
@@ -537,6 +538,75 @@ fn run() -> Result<()> {
...
+
+quick_main!(run);
+```
+
+The `std` library is leveraged to make a well formed IP/port with the
+[`SocketAddrV4`] and [`Ipv4Addr`] structs.  The loopback address is a special
+address that runs only on the local machine, and is available on all machines.
+
+By passing 0 to the [`TcpListener::bind`], the OS will assign an unused random
+port.  The address and port that the OS assigns is available using
```
Comment (budziq) [inline src/net.md]: Minor nit. Please do not mix tenses. Use present/past/future or optimally passive voice throughout the whole text

**Comment by budziq** [comment]
Ok drop the passive voice comments ;) https://github.com/brson/rust-cookbook/issues/139

File: src/net.md (hunk 22 above)
Comment (AndyGauge) [inline src/net.md]: OK, so I'm going to take out the details about what the tuple is and call it the request.  I think the links do a good job of explaining the details.

File: src/net.md (hunk 21 above)
Comment (AndyGauge) [inline src/net.md]: 'an incoming connection is waited synchronously by TcpListener::accept` <- another case where passive isn't the best.

Comment (budziq) [inline src/net.md]: yep

File: src/net.md (hunk 22 above)
Comment (AndyGauge) [inline src/net.md]: Does this work:

[`TcpListener::accept`] synchronously waits for an incoming connection and
returns a `(`[`TcpStream`],  [`SocketAddrV4`]`)` representing the request.
[`read_to_string`] will wait until the connection is closed, flushing the socket,
which can be tested with `telnet ip port`.  For example, if the program
shows Listening on 127.0.0.1:11500, run

**Comment by AndyGauge** [comment]
Do you want me to rebase?

File: src/net.md (hunk 22 above)
Comment (budziq) [inline src/net.md]: Looks excellent :+1:

**Comment by budziq** [comment]
Nice work! @AndyGauge 

LGTM @dtolnay

**Comment by AndyGauge** [comment]
Hold on, registered port reference exists.

**Comment by AndyGauge** [comment]
whew.  Thanks @budziq

**Comment by budziq** [comment]
No problem. Keep them coming @AndyGauge !

**Comment by dtolnay** [comment]
Great work.
